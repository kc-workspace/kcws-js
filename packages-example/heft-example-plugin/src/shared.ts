/**
 * Plugin name.
 *
 * @public
 */
export const PLUGIN_NAME = "example-plugin" as const;
