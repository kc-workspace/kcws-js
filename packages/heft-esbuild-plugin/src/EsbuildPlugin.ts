import type {
  HeftConfiguration,
  IHeftTaskPlugin,
  IHeftTaskSession,
} from "@rushstack/heft";

import { build } from "esbuild";

export interface IEsbuildOption {}

class EsbuildPlugin implements IHeftTaskPlugin<IEsbuildOption> {
  accessor?: object | undefined;

  apply(
    _session: IHeftTaskSession,
    _heftConfiguration: HeftConfiguration,
    _pluginOptions?: IEsbuildOption | undefined
  ): void {
    build({});

    throw new Error("Method not implemented.");
  }
}

export default EsbuildPlugin;
