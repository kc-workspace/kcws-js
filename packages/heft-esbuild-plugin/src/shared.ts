/**
 * Plugin name.
 *
 * @public
 */
export const PLUGIN_NAME = "esbuild-plugin" as const;

/**
 * The error message when user enabled watch mode.
 *
 * @internal
 */
export const UNSUPPORTED_WATCH_MODE =
  `Esbuild is only available when running without watch mode.` as const;
