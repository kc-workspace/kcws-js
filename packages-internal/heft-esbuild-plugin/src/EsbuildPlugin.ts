import type {
  HeftConfiguration,
  IHeftTaskPlugin,
  IHeftTaskSession,
} from "@rushstack/heft";

import { join } from "node:path";
import { existsSync } from "node:fs";
import { BuildOptions, build, version } from "esbuild";

import { PLUGIN_NAME, UNSUPPORTED_WATCH_MODE } from "./shared";

type EsBuildOptionWhitelist =
  | "bundle"
  | "minify"
  | "sourcemap"
  | "platform"
  | "target";

/**
 * The options for esbuild build() function.
 *
 * @public
 */
export interface IEsbuildOption
  extends Pick<BuildOptions, EsBuildOptionWhitelist> {
  /** The default entrypoint (default to main field on package.json file) */
  entrypoint?: string;
  /** The entrypoints list that overrides default value */
  entrypoints?: string[];
  /** Output directory name (default is 'lib-bundle') */
  output?: string;
}

class EsbuildPlugin implements IHeftTaskPlugin<IEsbuildOption> {
  private _accessor: undefined;

  public get accessor(): undefined {
    return this._accessor;
  }

  public apply(
    session: IHeftTaskSession,
    heftConfiguration: HeftConfiguration,
    options?: IEsbuildOption | undefined
  ): void {
    if (session.parameters.watch) {
      throw new Error(UNSUPPORTED_WATCH_MODE);
    }

    session.hooks.run.tapPromise(PLUGIN_NAME, async () => {
      session.logger.terminal.writeLine(`Using esbuild version ${version}`);

      const basePath = heftConfiguration.buildFolderPath;
      const entries = [];
      const length = options?.entrypoints?.length ?? 0;
      if (length > 0) {
        entries.push(
          ...options!
            .entrypoints!.map(p => {
              const output = p.replace("<pwd>", basePath);
              if (existsSync(output)) return output;
              session.logger.terminal.writeWarningLine(
                `Remove '${p}' entrypoint from list because file is missing`
              );
              return "";
            })
            .filter(p => p !== "")
        );
      } else {
        const main =
          heftConfiguration.projectPackageJson.main ??
          options?.entrypoint ??
          "";
        entries.push(join(basePath, main));
      }

      const output = options?.output ?? "lib-bundle";
      const outdir = join(basePath, output);
      const buildOptions: BuildOptions = {
        entryPoints: entries,
        outdir: outdir,
        bundle: options?.bundle ?? true,
        minify: options?.minify ?? true,
        sourcemap: options?.sourcemap ?? false,
      };

      if (options?.platform) buildOptions.platform = options.platform;
      if (options?.target) buildOptions.target = options.target;
      await build(buildOptions);
    });
  }
}

export default EsbuildPlugin;
