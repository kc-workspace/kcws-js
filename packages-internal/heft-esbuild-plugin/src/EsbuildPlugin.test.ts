jest.mock("esbuild");

import type {
  HeftConfiguration,
  IHeftParameters,
  IHeftTaskHooks,
  IHeftTaskRunHookOptions,
  IHeftTaskRunIncrementalHookOptions,
  IHeftTaskSession,
} from "@rushstack/heft";

import { AsyncParallelHook } from "tapable";
import { build } from "esbuild";
import { StringBufferTerminalProvider, Terminal } from "@rushstack/terminal";
import { MockScopedLogger } from "@rushstack/heft/lib/pluginFramework/logging/MockScopedLogger";

import EsbuildPlugin, { IEsbuildOption } from "./EsbuildPlugin";
import { UNSUPPORTED_WATCH_MODE } from "./shared";

const mockTaskSession = (
  parameters?: Partial<IHeftParameters>
): IHeftTaskSession => {
  const terminalProvider = new StringBufferTerminalProvider(false);
  const terminal = new Terminal(terminalProvider);
  const logger = new MockScopedLogger(terminal);

  const _parameters: IHeftParameters = Object.assign(
    { watch: false },
    parameters
  ) as IHeftParameters;

  return {
    logger,
    parameters: _parameters,
    hooks: {
      run: new AsyncParallelHook<IHeftTaskRunHookOptions>(),
      runIncremental:
        new AsyncParallelHook<IHeftTaskRunIncrementalHookOptions>(),
    } as IHeftTaskHooks,
    parsedCommandLine: undefined!,
    requestAccessToPluginByName: undefined!,
    taskName: "esbuild",
    tempFolderPath: "/tmp/test",
  };
};

const mockHeftConfiguration = (
  config?: Partial<HeftConfiguration>
): HeftConfiguration => {
  const base = {
    buildFolderPath: "/tmp/build",
    projectPackageJson: {
      name: "example",
      version: "v0.1.0",
    },
  } as HeftConfiguration;

  return Object.assign(base, config);
};

describe("Esbuild heft-plugin", () => {
  it("should create with parameterless constructor", () => {
    const plugin = new EsbuildPlugin();
    expect(plugin).not.toBeFalsy();
  });

  it("should return undefined as accessor", () => {
    const plugin = new EsbuildPlugin();
    expect(plugin.accessor).toBeUndefined();
  });

  it("should call build(<default>) when applied plugin", async () => {
    const plugin = new EsbuildPlugin();

    const session = mockTaskSession();
    const config = mockHeftConfiguration();
    const options: IEsbuildOption = {};

    plugin.apply(session, config, options);
    await session.hooks.run.promise(undefined!);

    expect(build).toHaveBeenCalledTimes(1);
    expect(build).toHaveBeenCalledWith({
      bundle: true,
      entryPoints: ["/tmp/build"],
      minify: true,
      outdir: "/tmp/build/lib-bundle",
      sourcemap: false,
    });
  });

  it("should call build() when applied plugin with main field", async () => {
    const plugin = new EsbuildPlugin();

    const session = mockTaskSession();
    const config = mockHeftConfiguration({
      projectPackageJson: {
        name: "example",
        version: "v1.0.0",
        main: "lib/start.js",
      },
    });
    const options: IEsbuildOption = {};

    plugin.apply(session, config, options);
    await session.hooks.run.promise(undefined!);

    expect(build).toHaveBeenCalledTimes(1);
    expect(build).toHaveBeenCalledWith({
      bundle: true,
      entryPoints: ["/tmp/build/lib/start.js"],
      minify: true,
      outdir: "/tmp/build/lib-bundle",
      sourcemap: false,
    });
  });

  it("should call build() when platform and target exist", async () => {
    const plugin = new EsbuildPlugin();

    const session = mockTaskSession();
    const config = mockHeftConfiguration();
    const options: IEsbuildOption = {
      platform: "browser",
      target: ["node12", "chrome"],
    };

    plugin.apply(session, config, options);
    await session.hooks.run.promise(undefined!);

    expect(build).toHaveBeenCalledTimes(1);
    expect(build).toHaveBeenCalledWith({
      bundle: true,
      entryPoints: ["/tmp/build"],
      minify: true,
      outdir: "/tmp/build/lib-bundle",
      sourcemap: false,
      platform: "browser",
      target: ["node12", "chrome"],
    });
  });

  it("should call build() when override entrypoint", async () => {
    const plugin = new EsbuildPlugin();

    const session = mockTaskSession();
    const config = mockHeftConfiguration();
    const options: IEsbuildOption = {
      entrypoint: "lib/index.js",
    };

    plugin.apply(session, config, options);
    await session.hooks.run.promise(undefined!);

    expect(build).toHaveBeenCalledTimes(1);
    expect(build).toHaveBeenCalledWith({
      bundle: true,
      entryPoints: ["/tmp/build/lib/index.js"],
      minify: true,
      outdir: "/tmp/build/lib-bundle",
      sourcemap: false,
    });
  });

  it("should call build() when override entrypoints", async () => {
    const plugin = new EsbuildPlugin();

    const session = mockTaskSession();
    const config = mockHeftConfiguration();
    const options: IEsbuildOption = {
      entrypoints: ["lib/index.js", "lib/start.js"],
    };

    plugin.apply(session, config, options);
    await session.hooks.run.promise(undefined!);

    expect(build).toHaveBeenCalledTimes(1);
    expect(build).toHaveBeenCalledWith({
      bundle: true,
      entryPoints: ["lib/index.js", "lib/start.js"],
      minify: true,
      outdir: "/tmp/build/lib-bundle",
      sourcemap: false,
    });
  });

  it("should error when watch mode enabled", async () => {
    const plugin = new EsbuildPlugin();

    const session = mockTaskSession({ watch: true });
    const config = mockHeftConfiguration();
    const options: IEsbuildOption = {};

    expect(() => {
      plugin.apply(session, config, options);
    }).toThrow(UNSUPPORTED_WATCH_MODE);
  });
});
