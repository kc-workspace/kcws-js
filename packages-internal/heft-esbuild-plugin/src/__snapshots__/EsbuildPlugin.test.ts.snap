// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Esbuild heft-plugin should call build() when override entrypoints 1`] = `"Remove 'lib/start.js' entrypoint from list because file is missing[n]"`;
